File: 0018-4sum/0018-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        
        int n = nums.length;
        
        HashSet<List<Integer>> ans = new HashSet<>();
        HashSet<Long> hashset;
        
        for(int i=0; i<n; i++){
    
            for(int j=i+1; j<n; j++){
                
                hashset = new HashSet<>();
                
                for(int k=j+1; k<n; k++){
                    
                    long sum = nums[i];
                    sum += nums[j];
                    sum += nums[k];
                    long fourth = target - sum;
                    
                    if(hashset.contains(fourth) == true){
                        
                        List<Integer> list = Arrays.asList(nums[i], nums[j], nums[k], (int)fourth);
                        Collections.sort(list);
                        
                        ans.add(list);
                    }
                    
                    hashset.add((long)nums[k]);
                }
            }
        }
        
        List<List<Integer>> result = new ArrayList<>(ans);
        return result;
        
        
    }
}

File: 0018-4sum/0018-4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        
        int n = nums.size();
        
        set<vector<int>> st;
        
        for(int i=0; i<n; i++){
            
            for(int j=i+1; j<n; j++){
                
                set<long long> s;
                for(int k=j+1; k<n; k++){
                    
                    long long sum = nums[i];
                    sum += nums[j];
                    sum += nums[k];
                     
                    long long fourth = target - sum;
                    
                    if(s.find(fourth) != s.end()){
                        
                        vector<int> t = {nums[i], nums[j], nums[k], (int)fourth };
                        sort(begin(t), end(t));
                        
                        st.insert(t);
                    }
                    
                    s.insert(nums[k]);
                }
            }
        }
        
        vector<vector<int>> res(begin(st), end(st));
        return res;
    }
};